app/services/twilio_api_service.rb

```diff
@@ -49,56 +56,55 @@ def test_client
     @test_client ||= Twilio::REST::Client.new(@twilio_test_account_sid, @twilio_test_auth_token)
   end
 
-  def send_sms(recipient_number, message, callback_url = nil)
+  def send_sms(recipient_number:, message:, callback_url: nil, context: {})
     sender_number = twilio_sender_sms_number
     recipient_number = parse_phone_number(recipient_number)
 
-    send_twilio_message(sender_number, recipient_number, message, callback_url)
+    send_twilio_message(sender_number, recipient_number, message, callback_url, context)
   end
 
-  def send_whatsapp(recipient_number, message, callback_url = nil)
+  def send_whatsapp(recipient_number:, message:, callback_url: nil, context: {})
     sender_number = "whatsapp:" + twilio_sender_whatsapp_number
     recipient_number = "whatsapp:" + parse_phone_number(recipient_number)
 
-    send_twilio_message(sender_number, recipient_number, message, callback_url)
+    send_twilio_message(sender_number, recipient_number, message, callback_url, context)
   end
 
   def parse_phone_number(number)
     parsed_number = Phonelib.parse(number, Rails.application.config.country[:abbreviation]).raw_national
     default_country_code + parsed_number
   end
 
-  def failed?
-    error.present?
-  end
-
   private
 
   def default_country_code
     Rails.application.config.country[:sms_country_code]
   end
 
-  def send_twilio_message(sender_number, recipient_number, message, callback_url = nil)
-    @response = client.messages.create(
+  def send_twilio_message(sender_number, recipient_number, message, callback_url, context)
+    client.messages.create(
       from: sender_number,
       to: recipient_number,
       status_callback: callback_url,
       body: message
     )
   rescue Twilio::REST::TwilioError => exception
-    @error = exception
-    report_error(exception)
+    # see the link above for a list of codes; the else case happens on network failures
+    if exception.respond_to?(:code)
```

Comment:
I looked through the error codes and we probably don't want to retry on any of them and probably also don't want to alert Sentry on any of them (maaaybe on one of them, but I'm not sure it's worth the extra code complexity).

Reply:
Let's put this knowledge into a comment here. eg. `Exceptions with codes are Twilio errors that will not succeed on retry. Log the error and swallow the exception.`

Reply:
I'll add this to a block comment at the top of the file along with the links.